### ISSUE TITLE
Panic during contextual validation of a possible side-chain block

### ISSUE THREAD

**Issue by rex4539** [ZcashFoundation/zebra#1464]
## Analysis

This panic happened when downloading the gossiped block 1066452 (00000000002cfbf02a2fb58073c76a823a6dedcd219f12a96b1da6544a9fb044), with the tip at 1066451.

But it's possible that there were multiple tips, and these blocks were on different side-chains.

We use `best_height_by_hash` to find the iterator length, but we should use `any_height_by_hash`.

## Logs

```
Dec 06 02:42:54.655  INFO {zebrad="5b687f87"}:sync: zebrad::components::sync: starting sync, obtaining new tips
Dec 06 02:42:54.657  INFO {zebrad="5b687f87"}:sync:obtain_tips:state: zebra_state::util: created block locator tip_height=Height(1066451) min_locator_height=1066352 locators=[Height(1066451), Height(1066450), Height(1066449), Height(1066447), Height(1066443), Height(1066435), Height(1066419), Height(1066387), Height(1066352)]
Dec 06 02:43:00.666  INFO {zebrad="5b687f87"}:sync: zebrad::components::sync: exhausted prospective tip set
Dec 06 02:43:00.666  INFO {zebrad="5b687f87"}:sync: zebrad::components::sync: waiting to restart sync timeout=45s
```

## Error
```
contextual validation requires at least 28 (POW_AVERAGING_WINDOW + POW_MEDIAN_BLOCK_SPAN) blocks
```

## Metadata
|key|value|
|--|--|
|**version**|3.0.0-alpha.0|
|**git commit**|5b687f87|
|**location**|zebra-state/src/service.rs:190:9|

## SpanTrace

<details>

```
SpanTrace:
   0: zebra_state::service::state
             at zebra-state/src/service.rs:564
   1: zebrad::components::inbound::downloads::download_and_verify
           with hash=00000000002cfbf02a2fb58073c76a823a6dedcd219f12a96b1da6544a9fb044
             at zebrad/src/components/inbound/downloads.rs:107
   2: zebrad::components::inbound::inbound
             at zebrad/src/components/inbound.rs:141
   3: zebra_network::peer::connection::msg_as_req
           with msg=inv
             at zebra-network/src/peer/connection.rs:564
   4: zebra_network::peer::handshake::peer
           with addr=143.110.216.2:8233
             at zebra-network/src/peer/handshake.rs:198
   5: zebrad::commands::
           with zebrad="5b687f87"
             at zebrad/src/commands.rs:56
```
</details>

## Backtrace

<details>

```
Backtrace:
   0: backtrace::backtrace::libunwind::trace
             at .cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.54/src/backtrace/libunwind.rs:90:5
      backtrace::backtrace::trace_unsynchronized
             at .cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.54/src/backtrace/mod.rs:66:5
   1: backtrace::backtrace::trace
             at .cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.54/src/backtrace/mod.rs:53:14
   2: backtrace::capture::Backtrace::create
             at .cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.54/src/capture.rs:176:9
   3: backtrace::capture::Backtrace::new
             at .cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.54/src/capture.rs:140:22
   4: color_eyre::config::PanicHook::panic_report
             at .cargo/registry/src/github.com-1ecc6299db9ec823/color-eyre-0.5.10/src/config.rs:982:18
   5: color_eyre::config::PanicHook::into_panic_hook::{{closure}}
             at .cargo/registry/src/github.com-1ecc6299db9ec823/color-eyre-0.5.10/src/config.rs:961:29
   6: std::panicking::rust_panic_with_hook
             at /rustc/1c389ffeff814726dec325f0f2b0c99107df2673/library/std/src/panicking.rs:595:17
   7: std::panicking::begin_panic::{{closure}}
             at .rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panicking.rs:520:9
   8: std::sys_common::backtrace::__rust_end_short_backtrace
             at .rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:141:18
   9: std::panicking::begin_panic
             at .rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panicking.rs:519:12

```
</details>

**Comment by teor2345** [comment]
Analysis:

This panic happened when downloading the gossiped block 1066452 (00000000002cfbf02a2fb58073c76a823a6dedcd219f12a96b1da6544a9fb044), with the tip at 1066451.

But it's possible that there were multiple tips, and these blocks were on different side-chains.

We use `best_height_by_hash` to find the iterator length, but we should use `any_height_by_hash`.

**Comment by teor2345** [comment]
Ambiguous method names are a root cause of this bug, we'll fix them in #1446.

**Comment by dconnolly** [comment]
Looks like this is still occuring despite the change in #1465 ?

## Error
```
contextual validation requires at least 28 (POW_AVERAGING_WINDOW + POW_MEDIAN_BLOCK_SPAN) blocks
```

## Metadata
|key|value|
|--|--|
|**version**|3.0.0-alpha.0|
|**git commit**|75b5e10|
|**location**|zebra-state/src/service.rs:190:9|


## SpanTrace

<details>

```
SpanTrace:
   0: zebra_state::service::state
             at zebra-state/src/service.rs:564
   1: zebrad::components::sync::obtain_tips
             at zebrad/src/components/sync.rs:255
   2: zebrad::components::sync::sync
             at zebrad/src/components/sync.rs:150
   3: zebrad::commands::
           with zebrad="75b5e10"
             at zebrad/src/commands.rs:56
```
</details>

## Backtrace

<details>

```
Backtrace:
   0: color_eyre::config::PanicHook::panic_report
   1: <zebrad::application::ZebradApp as abscissa_core::application::Application>::register_components::{{closure}}
   2: std::panicking::rust_panic_with_hook
             at rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:581:17
   3: std::panicking::begin_panic::{{closure}}
   4: std::sys_common::backtrace::__rust_end_short_backtrace
   5: std::panicking::begin_panic
   6: <zebra_state::service::StateService as tower_service::Service<zebra_state::request::Request>>::call
   7: <tower::util::boxed::sync::Boxed<S> as tower_service::Service<Request>>::call
   8: <tower::buffer::worker::Worker<T,Request> as core::future::future::Future>::poll
   9: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut

```
</details>

**Comment by teor2345** [comment]
Sentry is running commit 75b5e1081619592a76c52e1ec69330725ae4e31d, which forks from the main branch at #1458 
5b687f878ab6d602285510a468fd448f18916064, but the fix for this issue is in #1465

Please update Sentry to use the latest main branch

**Comment by dconnolly** [comment]
> Sentry is running commit [75b5e10](https://github.com/ZcashFoundation/zebra/commit/75b5e1081619592a76c52e1ec69330725ae4e31d), which forks from the main branch at #1458
> [5b687f8](https://github.com/ZcashFoundation/zebra/commit/5b687f878ab6d602285510a468fd448f18916064), but the fix for this issue is in #1465
> 
> Please update Sentry to use the latest main branch

Thanks! My bad.
